File: users-server/src/main/java/net/plshark/users/webservice/UsersController.java
package net.plshark.users.webservice;

import java.util.Objects;
import javax.validation.constraints.Min;
import net.plshark.ObjectNotFoundException;
import net.plshark.users.model.PasswordChangeRequest;
import net.plshark.users.model.User;
import net.plshark.users.model.UserInfo;
import net.plshark.users.service.UserManagementService;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * Controller providing web service methods for users
 */
@RestController
@RequestMapping("/users")
public class UsersController {

    private final UserManagementService userMgmtService;

    /**
     * Create a new instance
     * @param userManagementService the service to use to modify users
     */
    public UsersController(UserManagementService userManagementService) {
        this.userMgmtService = Objects.requireNonNull(userManagementService, "userManagementService cannot be null");
    }

    /**
     * Get all users up to the maximum result count
     * @param maxResults the maximum number of results to return
     * @param offset the offset to start the list at
     * @return the users
     */
    @GetMapping(produces = MediaType.APPLICATION_JSON_VALUE)
    public Flux<UserInfo> getAll(@RequestParam(value = "max-results", defaultValue = "50") @Min(1) int maxResults,
                                 @RequestParam(value = "offset", defaultValue = "0") @Min(0) long offset) {
        return userMgmtService.getAll(maxResults, offset)
                .map(UserInfo::fromUser);
    }

    /**
     * Retrieve a user by username
     * @param username the username
     * @return the matching user
     */
    @GetMapping(path = "/{username}", produces = MediaType.APPLICATION_JSON_VALUE)
    public Mono<UserInfo> getUser(@PathVariable("username") String username) {
        return userMgmtService.getUserByUsername(username)
                .map(UserInfo::fromUser)
                .switchIfEmpty(Mono.defer(() -> Mono.error(new ObjectNotFoundException("No user found for username"))));
    }

    /**
     * Insert a new user
     * @param user the user to insert
     * @return the inserted user
     */
    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
    public Mono<UserInfo> insert(@RequestBody User user) {
        return userMgmtService.insertUser(user)
            .map(UserInfo::fromUser);
    }

    /**
     * Delete a user by ID
     * @param id the user ID
     * @return an empty result
     */
    @DeleteMapping("/{id}")
    public Mono<Void> delete(@PathVariable("id") long id) {
        return userMgmtService.deleteUser(id);
    }

    /**
     * Change a user's password
     * @param userId the ID of the user
     * @param request the password change request
     * @return an empty result or ObjectNotFoundException if the user was not found or the current password was incorrect
     */
    @PostMapping(path = "/{id}/password")
    public Mono<Void> changePassword(@PathVariable("id") long userId, @RequestBody PasswordChangeRequest request) {
        return userMgmtService.updateUserPassword(userId, request.getCurrentPassword(), request.getNewPassword());
    }

    /**
     * Grant a role to a user
     * @param userId the ID of the user to grant to
     * @param roleId the ID of the role to grant
     * @return an empty result or ObjectNotFoundException if the user or role does not exist
     */
    @PostMapping(path = "/{userId}/roles/{roleId}")
    public Mono<Void> grantRole(@PathVariable("userId") long userId, @PathVariable("roleId") long roleId) {
        return userMgmtService.grantRoleToUser(userId, roleId);
    }

    /**
     * Remove a role from a user
     * @param userId the ID of the user to remove the role from
     * @param roleId the ID of the role to remove
     * @return an empty result or ObjectNotFoundException if the user does not exist
     */
    @DeleteMapping(path = "/{userId}/roles/{roleId}")
    public Mono<Void> removeRole(@PathVariable("userId") long userId, @PathVariable("roleId") long roleId) {
        return userMgmtService.removeRoleFromUser(userId, roleId);
    }
}

File: user-error-server/src/main/kotlin/net/plshark/users/webservice/UsersController.kt
package net.plshark.users.webservice

import net.plshark.errors.BadRequestException
import net.plshark.errors.DuplicateException
import net.plshark.errors.ObjectNotFoundException
import net.plshark.users.model.Group
import net.plshark.users.model.PasswordChangeRequest
import net.plshark.users.model.Role
import net.plshark.users.model.User
import net.plshark.users.model.UserCreate
import net.plshark.users.repo.UserGroupsRepository
import net.plshark.users.repo.UserRolesRepository
import net.plshark.users.repo.UsersRepository
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.dao.EmptyResultDataAccessException
import org.springframework.http.MediaType
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers
import javax.validation.constraints.Min

/**
 * Controller providing web service methods for users
 */
@RestController
@RequestMapping("/users")
class UsersController(
    private val userRepo: UsersRepository,
    private val userRolesRepo: UserRolesRepository,
    private val userGroupsRepo: UserGroupsRepository,
    private val passwordEncoder: PasswordEncoder
) {

    @GetMapping(produces = [MediaType.APPLICATION_JSON_VALUE])
    fun getUsers(
        @RequestParam(value = "max-results", defaultValue = "50") maxResults: @Min(1) Int,
        @RequestParam(value = "offset", defaultValue = "0") offset: @Min(0) Long
    ): Flux<User> {
        return userRepo.getAll(maxResults, offset)
    }

    @GetMapping(path = ["/{id}"], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun findById(@PathVariable("id") id: Long): Mono<User> {
        return userRepo.findById(id)
            .switchIfEmpty(Mono.error { ObjectNotFoundException("No user found for id") })
    }

    @PostMapping(consumes = [MediaType.APPLICATION_JSON_VALUE], produces = [MediaType.APPLICATION_JSON_VALUE])
    fun create(@RequestBody user: UserCreate): Mono<User> {
        if (user.password.isEmpty()) throw BadRequestException("Password cannot be empty")
        return Mono.just(user)
            .subscribeOn(Schedulers.parallel())
            .map { it.copy(password = passwordEncoder.encode(it.password)) }
            .flatMap { userRepo.insert(it) }
            .onErrorMap(DataIntegrityViolationException::class.java) { e: DataIntegrityViolationException? ->
                DuplicateException("A user with username ${user.username} already exists", e)
            }
    }

    @DeleteMapping("/{id}")
    fun delete(@PathVariable("id") id: Long): Mono<Void> {
        return userRepo.deleteById(id)
    }

    @PostMapping(path = ["/{id}/password"])
    fun changePassword(
        @PathVariable("id") id: Long,
        @RequestBody request: PasswordChangeRequest
    ): Mono<Void> {
        if (request.newPassword.isEmpty()) throw BadRequestException("New password cannot be empty")
        val newPasswordEncoded = passwordEncoder.encode(request.newPassword)
        val currentPasswordEncoded = passwordEncoder.encode(request.currentPassword)
        return findRequiredById(id)
            .flatMap { user: User ->
                userRepo.updatePassword(user.id, currentPasswordEncoded, newPasswordEncoded)
                    .onErrorResume(EmptyResultDataAccessException::class.java) { e: EmptyResultDataAccessException? ->
                        Mono.error(BadRequestException("Incorrect current password", e))
                    }
            }
    }

    @GetMapping(path = ["/{id}/roles"])
    fun getUserRoles(@PathVariable("id") id: Long): Flux<Role> {
        return userRolesRepo.findRolesByUserId(id)
    }

    @PutMapping(path = ["/{id}/roles"])
    @Transactional
    fun updateUserRoles(@PathVariable("id") id: Long, @RequestBody updatedRoles: Set<Role>): Flux<Role> {
        return findRequiredById(id)
            .thenMany(userRolesRepo.findRolesByUserId(id))
            .collectList()
            .map { existingRoles ->
                Flux.merge(
                    Flux.fromIterable(existingRoles.minus(updatedRoles))
                        .flatMap { userRolesRepo.deleteById(id, it.id) },
                    Flux.fromIterable(updatedRoles.minus(existingRoles))
                        .flatMap { userRolesRepo.insert(id, it.id) }
                )
            }.thenMany(Flux.fromIterable(updatedRoles))
    }

    @GetMapping(path = ["/{id}/groups"])
    fun getUserGroups(@PathVariable("id") id: Long): Flux<Group> {
        return userGroupsRepo.findGroupsByUserId(id)
    }

    @PutMapping(path = ["/{id}/groups"])
    @Transactional
    fun updateUserGroups(@PathVariable("id") id: Long, @RequestBody updatedGroups: Set<Group>): Flux<Group> {
        return findRequiredById(id)
            .thenMany(userGroupsRepo.findGroupsByUserId(id))
            .collectList()
            .map { existingGroups ->
                Flux.merge(
                    Flux.fromIterable(existingGroups.minus(updatedGroups))
                        .flatMap { userGroupsRepo.deleteById(id, it.id) },
                    Flux.fromIterable(updatedGroups.minus(existingGroups))
                        .flatMap { userGroupsRepo.insert(id, it.id) }
                )
            }.thenMany(Flux.fromIterable(updatedGroups))
    }

    private fun findRequiredById(id: Long): Mono<User> {
        return findById(id)
            .switchIfEmpty(Mono.error { ObjectNotFoundException("No user found for $id") })
    }
}
